### PR TITLE
Add fingerprinting for synchronous instruments

### LINKED ISSUE #221: Improve the fast path for synchronous instruments
**Describe the solution you'd like**
If the metric instrument and attributes have been used recently, there should be 0 allocations required to update the metric.

**Describe alternatives you've considered**
There is a widespread pattern for solving this, known as "fingerprinting". We will effectively hash the attribute set and perform a lookup before computing an `attribute.Set`.

**Additional context**
Both the Prometheus-Go library and an internal Statsd-like library used at Lightstep use the fingerprinting pattern to avoid allocations in the fast path for synchronous instruments.

Note that fingerprinting may lead to hash collisions, which for correctness have to be treated with care. This SDK is designed so that the reader code path is tolerant of duplicates in the update code path, which means that a fingerprint based on ordered, not-deduplicated attributes may be used correctly. Any duplication caused by ordering or duplicate key values will be eliminated in the read path.

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** 
This replaces an up-front `attribute.Set` construction (with compulsory allocs) into a fast path with no allocations.

```
name                               old time/op    new time/op    delta
CounterAddNoAttrs-10                 38.8ns ±11%    37.3ns ±21%     ~     (p=0.629 n=4+3)
CounterAddOneAttr-10                  170ns ± 0%      84ns ± 0%     ~     (p=0.057 n=4+3)
CounterAddManyAttrs-10               1.30µs ± 1%    1.11µs ± 5%     ~     (p=0.057 n=4+3)
CounterCollectOneAttrNoReuse-10       559ns ± 3%     473ns ± 1%     ~     (p=0.057 n=4+3)
CounterCollectOneAttrWithReuse-10     414ns ± 1%     327ns ± 1%     ~     (p=0.057 n=4+3)
CounterCollectTenAttrs-10            2.52µs ± 2%    1.65µs ± 0%     ~     (p=0.057 n=4+3)
CounterCollectTenAttrsTenTimes-10    25.1µs ± 1%    16.4µs ± 1%     ~     (p=0.057 n=4+3)

name                               old alloc/op   new alloc/op   delta
CounterAddNoAttrs-10                  0.00B          0.00B          ~     (all equal)
CounterAddOneAttr-10                   152B ± 0%       64B ± 0%  -57.89%  (p=0.000 n=4+3)
CounterAddManyAttrs-10                 543B ± 0%      593B ± 0%     ~     (p=0.057 n=4+3)
CounterCollectOneAttrNoReuse-10        488B ± 0%      400B ± 0%  -18.03%  (p=0.000 n=4+3)
CounterCollectOneAttrWithReuse-10      224B ± 0%      136B ± 0%  -39.29%  (p=0.000 n=4+3)
CounterCollectTenAttrs-10            1.59kB ± 0%    0.71kB ± 0%  -55.30%  (p=0.000 n=4+3)
CounterCollectTenAttrsTenTimes-10    15.9kB ± 0%     7.1kB ± 0%     ~     (p=0.057 n=4+3)

name                               old allocs/op  new allocs/op  delta
CounterAddNoAttrs-10                   0.00           0.00          ~     (all equal)
CounterAddOneAttr-10                   3.00 ± 0%      1.00 ± 0%  -66.67%  (p=0.000 n=4+3)
CounterAddManyAttrs-10                 6.00 ± 0%      7.00 ± 0%     ~     (p=0.057 n=4+3)
CounterCollectOneAttrNoReuse-10        9.00 ± 0%      7.00 ± 0%  -22.22%  (p=0.000 n=4+3)
CounterCollectOneAttrWithReuse-10      5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=4+3)
CounterCollectTenAttrs-10              32.0 ± 0%      12.0 ± 0%  -62.50%  (p=0.000 n=4+3)
CounterCollectTenAttrsTenTimes-10       320 ± 0%       120 ± 0%  -62.50%  (p=0.000 n=4+3)

```

**Link to tracking Issue:** 
Fixes #221.

**Testing:** 
Tests pass with and without a "real" fingerprint function, since collisions and duplicate-detection is implemented. One new tests covers hash collisions explicitly.

**Documentation:** No user-visible changes.

### CODE REVIEW

File: lightstep/sdk/metric/internal/syncstate/sync.go [hunk 1]
```diff
@@ -178,74 +222,179 @@ func capture[N number.Any, Traits number.Traits[N]](_ context.Context, inst *Ins
...
+		fp += fprint.Mix(
+			fprint.FingerprintString(string(attr.Key)),
+			fingerprintValue(attr.Value),
+		)
+	}
+
+	return fp
+}
+
+func fingerprintSlice[T any](slice []T, f func(T) uint64) uint64 {
```
Comment (jaronoff97) [inline lightstep/sdk/metric/internal/syncstate/sync.go:237]: 😍 generics are wonderful!

File: lightstep/sdk/metric/internal/syncstate/sync.go [hunk 2]
```diff
@@ -178,74 +222,179 @@ func capture[N number.Any, Traits number.Traits[N]](_ context.Context, inst *Ins
...
-	if rec, ok := inst.current[aset]; ok {
-		// Existing record case.
-		if rec.refMapped.ref() {
-			// At this moment it is guaranteed that the
-			// record is in the map and will not be removed.
-			return rec
-		}
+	rec := inst.current[fp]
+
+	// Note: we could (optionally) allow collisions and not scan this list.
```
Comment (jaronoff97) [inline lightstep/sdk/metric/internal/syncstate/sync.go:323]: Good note for a future configuration option :)

File: lightstep/sdk/metric/internal/syncstate/sync.go [hunk 3]
```diff
@@ -91,33 +92,67 @@ func (inst *Instrument) SnapshotAndProcess() {
...
-		// (b) if this is indeed the last reference, the collector needs the
-		// release signal.
-		_ = rec.conditionalSnapshotAndProcess(unmapped)
-
-		if unmapped {
-			// If any other goroutines are now trying to re-insert this
-			// entry in the map, they are busy calling Gosched() awaiting
-			// this deletion:
+
+		if keep == nil {
```
[the remark this answers is not in the document]
Comment (jmacd) [reply lightstep/sdk/metric/internal/syncstate/sync.go:114]: Sorry! I knew this function deserved comments and rushed it through. I've added a complete explanation.

File: lightstep/sdk/metric/internal/syncstate/sync.go (hunk 2 above)
Comment (jmacd) [reply lightstep/sdk/metric/internal/syncstate/sync.go:323]: I wonder how much it actually matters, looking forward to real-world benchmarks.

File: lightstep/sdk/metric/internal/syncstate/sync.go [hunk 4]
```diff
@@ -91,33 +92,67 @@ func (inst *Instrument) SnapshotAndProcess() {
...
 			delete(inst.current, key)
+			continue
+		}
+
+		last.next = nil
+
+		if keep != reclist {
+			inst.current[key] = keep
```
[the remark this answers is not in the document]
Comment (jmacd) [reply lightstep/sdk/metric/internal/syncstate/sync.go:122]: Done. To clarify this function, I renamed the two variables "head" and "tail". It's rebuilding a linked list.

**Comment by jmacd** [comment]
Added two commits: use a sync.Pool in once location, avoid aliasing a unnecessary slice
There are two more locations where a sync.Pool would help, but they're in generic code and Go generics don't help with generic package-scoped vars.

File: lightstep/sdk/metric/internal/syncstate/sync.go [hunk 5]
```diff
@@ -91,33 +98,77 @@ func (inst *Instrument) SnapshotAndProcess() {
...
+			if inst.singleSnapshotAndProcess(key, rec) {
+				if head == nil {
+					// The first time a record will be kept,
+					// it becomes the head and tail.
+					head = rec
+					tail = rec
+				} else {
+					// Subsequently, update the
+					// tail of the list.
+					tail.next = rec
```
Comment (jaronoff97) [inline lightstep/sdk/metric/internal/syncstate/sync.go:121]: Doesn't this mean that the tail record is pointing to itself? Should this be rec.next?

Comment (jaronoff97) [reply lightstep/sdk/metric/internal/syncstate/sync.go:121]: Or does the `tail.next = nil` on line 134 prevent this?

**Review by jaronoff97 — APPROVED** [review]
One non-blocking question to help me confirm my understanding of this approach

File: lightstep/sdk/metric/internal/syncstate/sync.go (hunk 5 above)
Comment (jmacd) [reply lightstep/sdk/metric/internal/syncstate/sync.go:121]: Yes. I added a note to clarify that the list becomes temporarily invalid.

**Pull request merged**